rfcs/20200920-keras-nlp-bert.md

```diff
@@ -0,0 +1,379 @@
+               inner_dropout=0.0,
+               attention_initializer=None,
+               **kwargs):
+    """Initializes `TransformerEncoderBlock`.
+
+    Arguments:
+      num_attention_heads: Number of attention heads.
+      inner_dim: The output dimension of the first Dense layer in a two-layer
+        feedforward network.
+      inner_activation: The activation for the first Dense layer in a two-layer
+        feedforward network.
+      output_range: the sequence output range, [0, output_range) for slicing the
+        target sequence. `None` means the target sequence is not sliced.
+      kernel_initializer: Initializer for dense layer kernels.
+      bias_initializer: Initializer for dense layer biases.
+      kernel_regularizer: Regularizer for dense layer kernels.
+      bias_regularizer: Regularizer for dense layer biases.
+      activity_regularizer: Regularizer for dense layer activity.
+      kernel_constraint: Constraint for dense layer kernels.
+      bias_constraint: Constraint for dense layer kernels.
+      use_bias: Whether to enable use_bias in attention layer. If set False,
+        use_bias in attention layer is disabled.
+      norm_first: Whether to normalize inputs to attention and intermediate
+        dense layers. If set False, output of attention and intermediate dense
+        layers is normalized.
+      norm_epsilon: Epsilon value to initialize normalization layers.
+      output_dropout: Dropout probability for the post-attention and output
+        dropout.
+      attention_dropout: Dropout probability for within the attention layer.
+      inner_dropout: Dropout probability for the first Dense layer in a
+        two-layer feedforward network.
+      attention_initializer: Initializer for kernels of attention layers. If set
+        `None`, attention layers use kernel_initializer as initializer for
+        kernel.
+      **kwargs: keyword arguments/
+    """
+```
+
+### Layers -- SelfAttentionMask
+
+```python
+class SelfAttentionMask(tf.keras.layers.Layer):
+  """Create 3D attention mask from a 2D tensor mask.
+
+    inputs[0]: from_tensor: 2D or 3D Tensor of shape
+      [batch_size, from_seq_length, ...].
+    inputs[1]: to_mask: int32 Tensor of shape [batch_size, to_seq_length].
+
+    Returns:
+      float Tensor of shape [batch_size, from_seq_length, to_seq_length].
+  """
+
+  def call(self, inputs, to_mask):
+```
+
+### Layers -- OnDeviceEmbedding
+
+```python
+class OnDeviceEmbedding(tf.keras.layers.Layer):
+  """Performs an embedding lookup suitable for accelerator devices.
+
+  This layer uses either tf.gather or tf.one_hot to translate integer indices to
+  float embeddings.
+
+  Arguments:
+    vocab_size: Number of elements in the vocabulary.
+    embedding_width: Output size of the embedding layer.
+    initializer: The initializer to use for the embedding weights. Defaults to
+      "glorot_uniform".
+    use_one_hot: Whether to use tf.one_hot over tf.gather for the embedding
```

Comment:
This class feels like it's leaking implementation details very high up the user interface. I understand why it's necessary now, but:
 1. in the future we might be able to optimize matmul(x, one_hot(y)) to gather(x, y) for the cases where it makes sense
 2. in the future we might want to decide whether embeddings should live on the host or on device holistically based on things like the embedding size

Given these two things I think it makes sense to file bugs for the infra folks to investigate making these tradeoffs automatically and to call this class experimental to indicate that we hope to simplify it in the future.

Reply:
Makes sense. I filed a bug. 
On the other hand, do you think it's more appropriate if I claim the ideal solution is that this would be part of tf.keras.layers.Embedding, so no embedding layers are required here?